### PR TITLE
Multisig spec

### PR DESCRIPTION
**Pull request by hansl**

### CODE REVIEW

File: src/ledger/src/multisig.rs
Comment (fmorency) [inline src/ledger/src/multisig.rs:40]: According to the spec, this could be any `multisig-transaction`.

File: src/ledger/src/multisig.rs
Comment (fmorency) [inline src/ledger/src/multisig.rs:151]: Remove `println!`

File: src/many-ledger/src/module.rs
Comment (fmorency) [inline src/many-ledger/src/module.rs:524]: I'll only comment it once, but I would like to see `pub type AddFeaturesReturn = EmptyReturn` and use that

File: src/ledger/src/multisig.rs
Comment (hansl) [reply src/ledger/src/multisig.rs:40]: The CLI will not support all the possible types because it becomes a bit insane (and recursive). If you have custom needs, you should use `many` instead.

File: src/many-ledger/src/module.rs
Comment (hansl) [reply src/many-ledger/src/module.rs:524]: will do.

File: src/ledger/src/multisig.rs
Comment (hansl) [reply src/ledger/src/multisig.rs:151]: See above

**Review by fmorency — CHANGES_REQUESTED** [review]
This branch doesn't build properly.

- `TransactionInfo::MultisigSubmit.memo` is still an `Option<ByteVec>`
- `TransactionInfo::MultisigSubmit.data` doesn't exists

Also, CICD to be fixed.

Finally, I'll wait on your feedback about `setDefaults` on the spec PR.

**Comment by hansl** [comment]
I still have to do the `data` change and the `SetDefaults` transaction type. I will also add a test having 2 accounts with multisig of multisig etc.

**Comment by fmorency** [comment]
@hansl I'll wait before reviewing this. Too many moving parts in the spec and in `many-rs`.

**Comment by hansl** [comment]
This should be ready to review, only thing missing is the other transactions support for multisig, which requires changes in many-rs and will be done in a separate PR. This is already very long.

File: src/many-ledger/src/storage.rs
Comment (fmorency) [inline src/many-ledger/src/storage.rs:806]: nit:

We need to test corner-cases such as

- A multisig transaction has been submitted. The list of approvers is X
- We remove the `canMultisigApprove` role from one of the user in X

Here, we never check that the sender has the `canMultisigApprove` role.

We should at least document the intended behavior.

File: src/many-ledger/src/storage.rs
Comment (fmorency) [inline src/many-ledger/src/storage.rs:52]: nit: I don't like the name; I feel like it's more of a `MultisigTransactionStorage`

**Review by fmorency — APPROVED** [review]
Thanks for this! Some comments.

As you mentioned, this is incomplete but already very big. I think we can merge this and make another iteration with bite-size PRs.

You already know this, but this feature needs many more tests. We should make a checklist somewhere of all the use-cases and corner-cases we need to test.

File: src/ledger/src/multisig.rs
Comment (hansl) [reply src/ledger/src/multisig.rs:62]: Done.

File: src/many-ledger/src/storage.rs
Comment (hansl) [reply src/many-ledger/src/storage.rs:806]: Okay I'll refactor here to be more in line with the spec. I think the current behaviour is unexpected yes.

**Comment by hansl** [comment]
Running CI then merging.

**Pull request merged**
